### PR TITLE
🚀 Release/1.2.0

### PR DESCRIPTION
**Pull request by skoruba**
# 🚀 Release/1.2.0
 
- Should the docker-compose.vs.*.yml reference .NET 6? - #54 
- Management for Identity Providers - #52 
- Enable Admin.UI StaticWebAssets in non-development environments as per Microsoft's documentation - #65 
- POST /api/ApiScopes successfully updated the database but returns routing error - #83 
- Upgrade to Duende.IdentityServer.* version 6.1.0 - #79

### CODE REVIEW

File: src/Skoruba.Duende.IdentityServer.Admin.UI/Resources/Areas/AdminUI/Views/IdentityProvider/IdentityProviderDelete.de.resx [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+    <value>System.Resources.ResXResourceWriter, System.Windows.Forms, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089</value>
+  </resheader>
+  <data name="ButtonDelete" xml:space="preserve">
+    <value>Identitäts-Provider löschen</value>
+  </data>
+  <data name="NavigationIdentityProviders" xml:space="preserve">
+    <value>Identitäts-Providers</value>
+  </data>
+  <data name="PageTitle" xml:space="preserve">
+    <value>Identitäts-Provider löschen</value>
```
Comment (hoba84) [inline src/Skoruba.Duende.IdentityServer.Admin.UI/Resources/Areas/AdminUI/Views/IdentityProvider/IdentityProviderDelete.de.resx:127]: Wrong translation, change to "Identitätsanbieter löschen"

File: src/Skoruba.Duende.IdentityServer.Admin.UI/Resources/Areas/AdminUI/Views/IdentityProvider/IdentityProviderDelete.de.resx [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+  </data>
+  <data name="PanelTitle" xml:space="preserve">
+    <value>Identitäts-Provider löschen</value>
```
Comment (hoba84) [inline src/Skoruba.Duende.IdentityServer.Admin.UI/Resources/Areas/AdminUI/Views/IdentityProvider/IdentityProviderDelete.de.resx:130]: Wrong translation, change to "Identitätsanbieter löschen"

File: src/Skoruba.Duende.IdentityServer.Admin.UI/Resources/Areas/AdminUI/Views/Shared/_Layout.de.resx [hunk 3]
```diff
@@ -141,6 +141,10 @@
   <data name="LinkIdentityResources" xml:space="preserve">
     <value>Identitäts-Ressources</value>
   </data>
+  <data name="LinkIdentityProviders" xml:space="preserve">
+    <value>Identity Providers</value>
+  </data>
+
   <data name="LinkLogout" xml:space="preserve">
     <value>Logout</value>
```
Comment (hoba84) [inline src/Skoruba.Duende.IdentityServer.Admin.UI/Resources/Areas/AdminUI/Views/Shared/_Layout.de.resx:149]: Wrong translation, change to "Abmelden"

File: src/Skoruba.Duende.IdentityServer.Admin.UI/Resources/Areas/AdminUI/Views/Shared/_Layout.de.resx [hunk 4]
```diff
@@ -141,6 +141,10 @@
   <data name="LinkIdentityResources" xml:space="preserve">
     <value>Identitäts-Ressources</value>
   </data>
+  <data name="LinkIdentityProviders" xml:space="preserve">
+    <value>Identity Providers</value>
```
Comment (hoba84) [inline src/Skoruba.Duende.IdentityServer.Admin.UI/Resources/Areas/AdminUI/Views/Shared/_Layout.de.resx:145]: Wrong translation, change to "Identitätsanbieter"

File: src/Skoruba.Duende.IdentityServer.Admin.UI/Resources/Areas/AdminUI/Views/Shared/_Layout.de.resx [hunk 5]
```diff
@@ -141,6 +141,10 @@
   <data name="LinkIdentityResources" xml:space="preserve">
     <value>Identitäts-Ressources</value>
```
Comment (hoba84) [inline src/Skoruba.Duende.IdentityServer.Admin.UI/Resources/Areas/AdminUI/Views/Shared/_Layout.de.resx:142]: Wrong translation, change to "Identitätsresourcen"

**Review by hoba84 — CHANGES_REQUESTED** [review]
Some review comments in exception handling (error scenarios) and EN, DE translation.

File: src/Skoruba.Duende.IdentityServer.Admin.Api/Controllers/IdentityProvidersController.cs [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+
+            return Ok();
+        }
+
+        [HttpDelete("{id}")]
+        public async Task<IActionResult> Delete(int id)
+        {
+            var identityProvider = new IdentityProviderDto { Id = id };
+
+            await _identityProviderService.GetIdentityProviderAsync(identityProvider.Id);
```
[the remark this answers is not in the document]
Comment (skoruba) [reply src/Skoruba.Duende.IdentityServer.Admin.Api/Controllers/IdentityProvidersController.cs:85]: There are validations in the Get method, if the ID exists.

**Comment by skoruba** [comment]
@hoba84 - thank you for your great feedback, can you please send a PR to this branch [release/1.2.0](https://github.com/skoruba/Duende.IdentityServer.Admin/tree/release/1.2.0) with your fixes regarding to translations etc? It would help a lot. Thanks!

File: src/Skoruba.Duende.IdentityServer.Admin.BusinessLogic/Services/IdentityProviderService.cs [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+            if (!canInsert)
+            {
+                throw new UserFriendlyViewException(string.Format(IdentityProviderServiceResources.IdentityProviderExistsValue().Description, identityProvider.Scheme), IdentityProviderServiceResources.IdentityProviderExistsKey().Description, identityProvider);
+            }
+
+            var entity = identityProvider.ToEntity();
+
+            var saved = await IdentityProviderRepository.AddIdentityProviderAsync(entity);
+
+            await AuditEventLogger.LogEventAsync(new IdentityProviderAddedEvent(identityProvider));
```
[the remark this answers is not in the document]
Comment (rpbrongers) [reply src/Skoruba.Duende.IdentityServer.Admin.BusinessLogic/Services/IdentityProviderService.cs:79]: The code follows the exact pattern of all the other services- I figured we don't want to just have one deviating from all the others. So perhaps we should 'upgrade' the comment to a backlog item that applies to the entire code base, not just this file.

File: src/Skoruba.Duende.IdentityServer.Admin.Api/Controllers/IdentityProvidersController.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (rpbrongers) [reply src/Skoruba.Duende.IdentityServer.Admin.Api/Controllers/IdentityProvidersController.cs:85]: The code follows the exact pattern of all the other controllers - I figured we don't want to just have one deviating from all the others. So perhaps we should 'upgrade' the comment to a backlog item that applies to the entire code base, not just this file.

File: src/Skoruba.Duende.IdentityServer.Admin.UI/Areas/AdminUI/Controllers/IdentityProviderController.cs [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+            // also: we check for duplicate keys:
+            var dupGroups = identityProvider.Properties.GroupBy(s => s.Value.Name).Where(g => g.Count() > 1).ToArray();
+            if (dupGroups.Length > 0)
+            {
+                foreach (var dupGroup in dupGroups)
+                {
+                    foreach (var d in dupGroup)
+                    {
+                        //duplicate key = not allowed
+                        ModelState.AddModelError($"{nameof(IdentityProviderDto.Properties)}[{d.Key}].{nameof(IdentityProviderPropertyDto.Name)}", $"duplicate name ({d.Value.Name})");
```
[the remark this answers is not in the document]
Comment (rpbrongers) [reply src/Skoruba.Duende.IdentityServer.Admin.UI/Areas/AdminUI/Controllers/IdentityProviderController.cs:72]: Why? No other model validation errors do that.

File: src/Skoruba.Duende.IdentityServer.Admin.EntityFramework/Repositories/IdentityProviderRepository.cs [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+                var existsWithSameName = await DbContext.IdentityProviders.Where(x => x.Scheme == identityProvider.Scheme && x.Id != identityProvider.Id).SingleOrDefaultAsync();
+                return existsWithSameName == null;
+            }
+        }
+        
+        public virtual async Task<int> DeleteIdentityProviderAsync(IdentityProvider identityProvider)
+        {
+            var identityProviderToDelete = await DbContext.IdentityProviders.Where(x => x.Id == identityProvider.Id).SingleOrDefaultAsync();
+
+            DbContext.IdentityProviders.Remove(identityProviderToDelete);
```
[the remark this answers is not in the document]
Comment (rpbrongers) [reply src/Skoruba.Duende.IdentityServer.Admin.EntityFramework/Repositories/IdentityProviderRepository.cs:85]: The code follows the exact pattern of all the other repositories - I figured we don't want to just have one deviating from all the others. So perhaps we should 'upgrade' the comment to a backlog item that applies to the entire code base, not just this file.

**Pull request closed without merging**